hpen.c

```diff
@@ -0,0 +1,456 @@
+		.name = "BARL",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_BARREL_SWITCH),
+		.code = BTN_STYLUS,
+		.required = false,
+	},
+	[HPEN_INVERT] = {
+		.name = "INVR",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_INVERT),
+		.code = BTN_TOOL_RUBBER,
+		.required = true,
+	},
+	[HPEN_ERASER] = {
+		.name = "ERSR",
+		.usage = HID_USAGE2(HUP_DIGITIZERS, HUD_ERASER),
+		.code = BTN_TOUCH,
+		.required = true,
+	},
+};
+
+
+struct hpen_absinfo {
+	int32_t min;
+	int32_t max;
+	int32_t res;
+};
+
+#define	USAGE_SUPPORTED(caps, usage)	bit_test(caps, usage)
+#define	HPEN_FOREACH_USAGE_ABS(caps, usage)			\
+	for ((usage) = 0; (usage) < HPEN_N_USAGES_ABS; ++(usage))	\
+		if (USAGE_SUPPORTED((caps), (usage)))
+#define	HPEN_FOREACH_USAGE_KEY(caps, usage)			\
+	for ((usage) = 0; (usage) < HPEN_N_USAGES_KEY; ++(usage))	\
+		if (USAGE_SUPPORTED((caps), (usage)))
+
+
+struct hpen_softc {
+	device_t sc_dev;
+
+	struct evdev_dev *sc_evdev;
+
+	uint8_t                 report_id;
+	bitstr_t                bit_decl(abs_caps, HPEN_N_USAGES_ABS);
+	bitstr_t                bit_decl(key_caps, HPEN_N_USAGES_KEY);
+	uint32_t                isize;
+	struct hpen_absinfo     ai[HPEN_N_USAGES_ABS];
+	struct hid_location     locs_abs[HPEN_N_USAGES_ABS];
+	struct hid_location     locs_key[HPEN_N_USAGES_KEY];
+};
+
+
+static device_probe_t hpen_probe;
+static device_attach_t hpen_attach;
+static device_detach_t hpen_detach;
+
+static evdev_open_t hpen_ev_open;
+static evdev_close_t hpen_ev_close;
+
+static const struct evdev_methods hpen_evdev_methods = {
+	.ev_open = &hpen_ev_open,
+	.ev_close = &hpen_ev_close,
+};
+
+static const struct hid_device_id hpen_devs[] = {
+	{ HID_TLC(HUP_DIGITIZERS, HUD_PEN) },
+};
+
+static int
+hpen_ev_close(struct evdev_dev *evdev)
+{
+	device_t dev = evdev_get_softc(evdev);
+
+	mtx_assert(hidbus_get_lock(dev), MA_OWNED);
+
+	return (hidbus_set_xfer(dev, 0));
+}
+
+static int
+hpen_ev_open(struct evdev_dev *evdev)
+{
+	device_t dev = evdev_get_softc(evdev);
+
+	mtx_assert(hidbus_get_lock(dev), MA_OWNED);
+
+	return (hidbus_set_xfer(dev, HID_XFER_READ));
+}
+
+static void
+hpen_intr(void *context, void *buf, uint16_t len)
+{
+	device_t dev = context;
+	struct hpen_softc *sc = device_get_softc(dev);
+	size_t usage;
+	uint8_t id;
+
+	mtx_assert(hidbus_get_lock(dev), MA_OWNED);
+
+	/* Ignore irrelevant reports */
+	id = sc->report_id != 0 ? *(uint8_t *)buf : 0;
+	if (sc->report_id != id) {
+		DPRINTF("Skip report with unexpected ID: %hhu\n", id);
+		return;
+	}
+
+	/* Make sure we don't process old data */
+	if (len < sc->isize)
+		bzero((uint8_t *)buf + len, sc->isize - len);
+
+	/* Strip leading "report ID" byte */
+	if (sc->report_id) {
+		len--;
+		buf = (uint8_t *)buf + 1;
+	}
+
+	HPEN_FOREACH_USAGE_ABS(sc->abs_caps, usage) {
+		if (sc->locs_abs[usage].size > 0 && hpen_hid_map_abs[usage].code != HPEN_NO_CODE)
+			evdev_push_abs(sc->sc_evdev, hpen_hid_map_abs[usage].code,
+					hid_get_data(buf, len, &sc->locs_abs[usage]));
```

Comment:
It is better to use unsigned version of hid_get_data (hid_get_udata) here to not be beaten by sign extension effects. For example, signed bool true == -1 not just 1